### PR TITLE
[Spec][PD] Enable fused TopK for GLM-5.2 MTP IndexShare

### PR DESCRIPTION
**Pull request by HanHan009527**
## Motivation

GLM-5.2 MTP reuses DSA IndexShare top-k seeds across draft-decode steps to avoid redundant indexer work. The seed-transfer path added in [PR #30839](https://github.com/sgl-project/sglang/pull/30839) preserves allocator independence by transporting request-relative positions. Because the existing fused TopK consumes allocator-local physical KV slots, this path currently falls back to an unfused page-table transform during draft decode.

This PR optimizes GLM-5.2 MTP IndexShare specifically under PD disaggregation. It converts the transferred seed once at the Decode consumer and then uses the existing fused TopK. The wire representation remains unchanged, while the Decode-local remap is required because the Prefill and Decode allocators are independent.

## Modifications

- Keep GLM-5.2 MTP IndexShare seeds in request-relative position space during transfer.
- For the supported Decode consumer path (`SGLANG_DSA_FUSE_TOPK=1`, non-HiSparse, and `dcp_size == 1`), remap the seed once through the local `req_to_token` table before it enters the EAGLE draft loop.
- Fail invalid seed rows closed instead of consuming stale, out-of-range, or unallocated slots.
- Enable the existing fused DSA TopK after the local remap.
- Add one focused CPU unit test covering fused-path selection and position-to-local-slot remapping.

## Accuracy Tests

The real PD seed was produced on Prefill, transferred to Decode, remapped locally, and consumed by fused TopK.

- MTP acceptance: `66% -> 66%`
- Decoded tokens per iteration: `2.96 -> 2.96`
- No request errors, OOMs, worker restarts, or long-context CUDA Graph fallback
- Targeted CPU suite: `19 passed, 17 warnings, 2 subtests passed`

A bounded LongBench-v2 `short` evaluation used the same first 64 sample IDs in both independent distributed runs, with temperature 0, `reasoning_effort=none`, `max_tokens=1024`, and concurrency 1. This is a correctness/runtime guard, not a full LongBench-v2 score.

| Metric | fused off | fused on |
| --- | ---: | ---: |
| Accuracy | 60.94% (39/64) | 62.50% (40/64) |
| Invalid answers | 1 | 2 |
| Request errors | 0 | 0 |
| Natural / max-token stops | 63 / 1 | 62 / 2 |
| Input tokens, p50 / p90 / max | 24584 / 47323 / 76495 | 24584 / 47323 / 76495 |

Answer choices agreed on `51/64` samples. Independent distributed runs are not bitwise deterministic, so aggregate accuracy and request/runtime guardrails are used for the regression check rather than exact text equality.

## Speed Tests and Profiling

Hardware and topology:

- 4 nodes, 32 x H20-96G
- Prefill: 2 nodes, TP16, EAGLE/MTP
- Decode: 2 nodes, TP16, EAGLE/MTP, Full CUDA Graph
- HiSparse disabled

Workload:

- 131072 input tokens + 1536 output tokens
- Concurrency 1, temperature 0, fixed seed
- One excluded warmup followed by three independent cache-miss runs per arm

| Metric | fused off | fused on | Delta |
| --- | ---: | ---: | ---: |
| TPOT | 11.60 ms | 11.22 ms | **-3.28%** |
| ITL | 34.37 ms | 33.31 ms | **-3.08%** |
| Decode forward | 17828.26 ms | 17274.41 ms | **-3.11%** |
| Output throughput | 26.23 tok/s | 26.52 tok/s | **+1.11%** |
| MTP acceptance | 66% | 66% | unchanged |
| Decoded tokens / iteration | 2.96 | 2.96 | unchanged |

TTFT and PD transfer time were unchanged within noise, as expected for a Decode-local optimization. Target verify, draft extend, and draft decode remained on Full CUDA Graph in all measured runs.

### Dataset-derived Decode operator profile

One representative request from the same bounded LongBench-v2 subset was profiled through Router -> Prefill -> Decode: sample index 24, 76,489 dataset-prompt tokens and 76,503 API prompt tokens after chat wrapping. The capture covered 32 active Decode iterations with Full CUDA Graph enabled. Only Decode TP0 is summarized below; all eight rank traces per arm were checked and retained for audit.

| Operator (Decode TP0) | fused off | fused on |
| --- | ---: | ---: |
| `transform_index_page_table_prefill_kernel` | 2528 calls | **0 calls** |
| `transform_index_page_table_decode_kernel` | 32 calls | **0 calls** |
| Unfused `topk_kernel` | 704 calls | **0 calls** |
| Fused `topk_small_batch_kernel<true>` | 0 calls | **704 calls** |
| `topk_plan` | 96 calls | 96 calls |
| FP8 paged-MQA logits | 704 calls | 704 calls |

This isolates the intended mechanism: recurring page-table conversion disappears, unfused TopK dispatch changes to the existing fused small-batch path, and indexer scoring/planning work is preserved. Raw profiler durations are intentionally not used for an A/B latency claim because capture perturbation differs between the two CUDA Graph paths; the profiler-off repeated `128K + 1.5K` A/B above remains the performance evidence.

## Checklist

- [x] The change is limited to the PD Decode DSA IndexShare path.
- [x] Formatting and repository pre-commit checks pass.
- [x] A focused unit test covers the new remap behavior.
- [x] Accuracy and real-hardware performance results are provided.
- [x] Dataset-derived Decode profiling confirms the recurring page-table transforms disappear.
- [x] No user-facing documentation change is required.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #30874159469](https://github.com/sgl-project/sglang/actions/runs/30874159469)
Latest PR Test (Extra): :white_check_mark: [Run #30874159192](https://github.com/sgl-project/sglang/actions/runs/30874159192)

### CODE REVIEW

**Comment by HanHan009527** [comment]
@zRzRzRzRzRzRzR Hello, when you have time, could you please take a look and see if this modification matches your design?

**Review by zRzRzRzRzRzRzR — COMMENTED** [review]
Just to confirm, this only changes the PD Decode path with fused TopK enabled, HiSparse disabled, and dcp_size == 1, while all existing paths retain their current behavior, correct?

**Review by zRzRzRzRzRzRzR — APPROVED** [review]

**Comment by zhao18235158400** [comment]
Hi @HanHan009527 — I read your PD/MTP IndexShare optimization where request-relative seeds are remapped once to decode-local KV slots, eliminating the repeated page-table transforms and 704 unfused TopK calls in the profiled run. I’m Daniel, a user researcher with a product research team studying local AI systems for inference and kernel engineers. My main question is: what observability would make allocator-identity bugs across prefill, transfer, and decode easier to isolate without a large distributed reproduction? Would you be open to a Zoom interview of up to 30 minutes at your convenience? We’re preparing a prototype and may later provide test units or invite relevant participants as early users or technical advisors. This is research, not sales.

**Comment by HanHan009527** [comment]
> Just to confirm, this only changes the PD Decode path with fused TopK enabled, HiSparse disabled, and dcp_size == 1, while all existing paths retain their current behavior, correct?

Yes， I ensured this in `should_remap_pd_dsa_seed_to_local_slots`.
```
def should_remap_pd_dsa_seed_to_local_slots(server_args: "ServerArgs") -> bool:
    """Whether a PD seed should enter the allocator-local fused TopK domain."""
    return (
        envs.SGLANG_DSA_FUSE_TOPK.get()
        and server_args.disaggregation_mode == "decode"
        and not server_args.enable_hisparse
        and server_args.dcp_size == 1
    )

``` 

HiSparse consumes logical/request-relative positions and resolves GPU residency per layer through `swap_in_selected_pages`; eagerly remapping the seed through `req_to_token` would produce logical or stale/nonresident slots and bypass the required host-to-device resolution. 

DCP shards KV across context ranks, while the current remap only produces a single-rank allocator-local slot and does not apply the DCP owner filter or global-to-local index conversion.

**Comment by HanHan009527** [comment]
@Fridge003 Could you please review this PR?

File: python/sglang/srt/layers/attention/dsa/utils.py [hunk 1]
```diff
@@ -58,21 +58,32 @@ def aiter_can_use_preshuffle_paged_mqa() -> bool:
...
+    from sglang.srt.server_args import ServerArgs
 
 
 def compute_dsa_seqlens(original_seq_lens, dsa_index_topk: int):
     return original_seq_lens.clamp(max=dsa_index_topk)
 
 
+def should_remap_pd_dsa_seed_to_local_slots(server_args: "ServerArgs") -> bool:
+    """Whether a PD seed should enter the allocator-local fused TopK domain."""
+    return (
```
Comment (kpham-sgl) [inline python/sglang/srt/layers/attention/dsa/utils.py:76]: guard this with is_cuda() because it seems like the NPU path has slightly different implementation
https://github.com/sgl-project/sglang/blob/bae8eb8d6caa26a8820ce70c35cdfe57fa5182b3/python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py#L428-L439
https://github.com/sgl-project/sglang/blob/bae8eb8d6caa26a8820ce70c35cdfe57fa5182b3/python/sglang/srt/hardware_backend/npu/graph_runner/eagle_draft_npu_graph_runner.py#L62-L85

File: python/sglang/srt/layers/attention/dsa/utils.py [hunk 2]
```diff
@@ -58,21 +58,32 @@ def aiter_can_use_preshuffle_paged_mqa() -> bool:
...
+        envs.SGLANG_DSA_FUSE_TOPK.get()
+        and server_args.disaggregation_mode == "decode"
+        and not server_args.enable_hisparse
+        and server_args.dcp_size == 1
+    )
+
+
 def should_use_dsa_fused_topk(
```
Comment (kpham-sgl) [inline python/sglang/srt/layers/attention/dsa/utils.py:85]: This behavior has become quite complicated. We should document somewhere that 

For prefill worker
- Target prefill: fused topk enabled
- Draft extend: fused topk disabled

For draft worker
- Draft extend, target verify, draft decode all have fused topk enabled after this PR

File: python/sglang/srt/speculative/eagle_disaggregation.py [hunk 3]
```diff
@@ -55,6 +58,32 @@ def build_eagle_disagg_draft_input(
     dsa_indices_list = [req.output_dsa_topk_indices for req in batch.reqs]
     if dsa_indices_list and all(t is not None for t in dsa_indices_list):
         dsa_topk_indices = torch.stack(dsa_indices_list, dim=0).to(batch.device)
+        if should_remap_pd_dsa_seed_to_local_slots(server_args):
+            # PD transports request-relative positions because the prefill and
+            # decode allocators are independent. The fused TopK path consumes
+            # physical slots, so materialize them once through the decode-local
+            # page table before the seed enters the draft loop/CUDA graph.
```
Comment (kpham-sgl) [inline python/sglang/srt/speculative/eagle_disaggregation.py]: Trim this comment

**Review by kpham-sgl — COMMENTED** [review]
Thank you for the contribution! Open question: how do we write better tests (outside of `test/registered/models_e2e/test_dsa_glm52_cache_layer_split.py`)

**Comment by kpham-sgl** [comment]
/rerun-test test/registered/models_e2e/test_dsa_glm52_cache_layer_split.py

File: python/sglang/srt/layers/attention/dsa/utils.py (hunk 1 above)
Comment (HanHan009527) [reply python/sglang/srt/layers/attention/dsa/utils.py:76]: done

File: python/sglang/srt/layers/attention/dsa/utils.py (hunk 2 above)
Comment (HanHan009527) [reply python/sglang/srt/layers/attention/dsa/utils.py:85]: done

**Comment by HanHan009527** [comment]
/rerun-test test/registered/models_e2e/test_dsa_glm52_cache_layer_split.py

**Comment by HanHan009527** [comment]
> Thank you for the contribution! Open question: how do we write better tests (outside of `test/registered/models_e2e/test_dsa_glm52_cache_layer_split.py`)

The review feedback has been processed. Regarding the testing issue, I think we might need to build a tiny model to test with 2 or 4 GPUs？

**Review by kpham-sgl — APPROVED** [review]
I think we can revisit the testing question later

**Comment by Fridge003** [comment]
/rerun-test test/registered/models_e2e/test_dsa_glm52_cache_layer_split.py

**Comment by Fridge003** [comment]
/rerun-test test/registered/models_e2e/test_dsa_glm52_cache_layer_split.py

**Comment by tianxiaojiang4** [comment]
Fused DSA top-k is still off on ROCm under PD: `should_remap_pd_dsa_seed_to_local_slots()` gates on `is_cuda()`, so this PR's remap never applies on gfx950, `pd_index_share_seed` stays true, and `should_use_dsa_fused_topk()` returns False.

Measured on GLM-5.2-MXFP4, 8×MI355X (gfx950, ROCm 7.2.0), 1P1D at TP2, decode step time p50:

| | |
|---|---|
| fused | **23.2 ms** |
| unfused (current ROCm behaviour) | **59.2 ms** |

2.55×, replicated across four independent run pairs. The gap is flat in context (39.7 ms at 37k prompt tokens, 41.8 ms at 177k) and both arms replay the draft CUDA graph (eager fraction 2e-4), so it is kernel work rather than launch overhead or a fallback.

Fusion itself is fine on gfx950. Enabling it by declining the seed passes needle-in-a-haystack at ~150k context (3/3 depths) with real acceptance 3.09, acceptance simulation off.

Is the `is_cuda()` gate intentional, or just untested on ROCm? Happy to test a patch on MI355X.

**Comment by kpham-sgl** [comment]
@tianxiaojiang4 , I meant to gate off NPU only but ROCm paths are untested. Can you submit a PR with tested results?
